File: leetcode-js/124-binary-tree-maximum-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
    if (!root) {
        return 0;
    }
    const [max] = _maxPathSum(root, -Infinity);
    return max;
};

function _maxPathSum(root, max) {
    if (!root) {
        return [max, 0];
    }

    let leftSum;
    let rightSum;

    [max, leftSum] = _maxPathSum(root.left, max);
    [max, rightSum] = _maxPathSum(root.right, max);

    const current = Math.max(root.val, root.val + leftSum, root.val + rightSum);
    max = Math.max(max, current, root.val + leftSum + rightSum);

    return [max, current];
}

File: leetcode-java/124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int maxPathSum(TreeNode root) {
        int[] max = new int[1];
        max[0] = Integer.MIN_VALUE;

        helper(root, max);

        return max[0];
    }

    int helper(TreeNode root, int[] max) {
        if (root == null) {
            return 0;
        }

        int left = helper(root.left, max);
        int right = helper(root.right, max);

        int currentMax = Math.max(Math.max(root.val, root.val + left), root.val + right);
        max[0] = Math.max(Math.max(max[0], currentMax), root.val + left + right);

        return currentMax;
    }
}
